sktime/transformations/base.py

```diff
@@ -549,9 +549,42 @@ def fit_transform(self, X, y=None):
                 then the return is a `Panel` object of type `pd-multiindex`
                 Example: i-th instance of the output is the i-th window running over `X`
         """
-        # Non-optimized default implementation; override when a better
-        # method is possible for a given algorithm.
-        return self.fit(X, y).transform(X, y)
+        # if fit is called, estimator is reset, including fitted state
+        self.reset()
```

Comment:
As these checks re-occur in fit and transform, I believe it would be better to refactor everything into one method and call it from fit, transform, fit_transform rather than copying all the content.

Reply:
fit and transform have a separate set of checks and datatype changes. I considered making methods like `_transform_checks` `_fit_checks` but it's not really that compatible. i.e. we don't need `self.check_is_fitted()` or multiple calls of `self._check_X_y`.

Reply:
That is fine to have separate checks in `transform` and `fit`. But it feels bad to copy these to `fit_transform`. Would parametrization work?

Reply:
The vast majority of the actual functionality is already contained in separate methods i.e. `self._check_X_y`, `update_data`, `self._convert_output`. I don't really feel so bad about duplicating calls to these. What remains is stuff like         
```
if self.get_tag("requires_y") and y is None:
    raise ValueError(f"{self.__class__.__name__} requires `y` in `fit`.")
```
which is small. The overall method is less than 30 lines including comments.

Reply:
Yes, but the code already now diverges. And at least to me it is not clear why.

Reply:
My point still being: every future change to fit or transform will have to be added to fit_transform, too. This is always a potential source of error. Thus, I am in favor of moving as much as possible into the same functionality.

Reply:
Fair. I can move these into separate methods
```
        # convert to output mtype
        if not hasattr(self, "_output_convert") or self._output_convert == "auto":
            X_out = self._convert_output(Xt, metadata=metadata)
        else:
            X_out = Xt
```
```
        # if fit is called, estimator is reset, including fitted state
        self.reset()

        # if requires_y is set, y is required in fit and update
        if self.get_tag("requires_y") and y is None:
            raise ValueError(f"{self.__class__.__name__} requires `y` in `fit`.")
```

or do you want the whole block from fit posted below in a single method with parameters?

Reply:
Does this work / make sense?

```python
    def _fit_checks(self, X, y, return_metadata=False):
        self.reset()

        # skip everything if fit_is_empty is True and we do not need to remember data
        if self.get_tag("fit_is_empty") and not self.get_tag("remember_data", False):
            self._is_fitted = True
            return True, None, None

        # if requires_y is set, y is required in fit and update
        if self.get_tag("requires_y") and y is None:
            raise ValueError(f"{self.__class__.__name__} requires `y` in `fit`.")

        # check and convert X/y
        if return_metadata:
            X_inner, y_inner, metadata = self._check_X_y(X=X, y=y, return_metadata=True)
        else:
            X_inner, y_inner = self._check_X_y(X=X, y=y)
            metadata = None

        # memorize X as self._X, if remember_data tag is set to True
        if self.get_tag("remember_data", False):
            self._X = update_data(None, X_new=X_inner)

        # skip the rest if fit_is_empty is True
        if self.get_tag("fit_is_empty"):
            self._is_fitted = True
            return True, None, None

        if return_metadata:
            return False, X_inner, y_inner, metadata
        else:
            return False, None, None
            
```


and in `fit` call:
```python
     def fit(self, X, y=None):
        fit_is_empty, X_inner, y_inner = self._fit_checks(X, y)
        if fit_is_empty:
            return self
```

and in `fit_transform` call:
```python
     def fit(self, X, y=None):
        _, X_inner, y_inner, metadata = self._fit_checks(X, y, True)
```

Reply:
That will skip checks and return nothing for `fit_transform` if `self.get_tag("fit_is_empty")`. A bunch of the fit checks are early abandon, but we can't do that in `fit_transform`

Reply:
Thanks. True. Another try:

```python
    def _fit_checks(self, X, y, early_abandon=True, return_metadata=False):
        self.reset()

        X_inner = None
        y_inner = None
        metadata = None

        # skip everything if fit_is_empty is True and we do not need to remember data
        if not early_abandon or \
                not self.get_tag("fit_is_empty") or \
                self.get_tag("remember_data", False):
            # if requires_y is set, y is required in fit and update
            if self.get_tag("requires_y") and y is None:
                raise ValueError(f"{self.__class__.__name__} requires `y` in `fit`.")

            # check and convert X/y
            X_inner, y_inner, metadata = self._check_X_y(X=X, y=y, return_metadata=True)

            # memorize X as self._X, if remember_data tag is set to True
            if self.get_tag("remember_data", False):
                self._X = update_data(None, X_new=X_inner)

        self._is_fitted = True

        if return_metadata:
            return X_inner, y_inner, metadata
        else:
            return X_inner, y_inner
```



and in `fit` call:
```python
     def fit(self, X, y=None):
        X_inner, y_inner = self._fit_checks(X, y)
        if self.get_tag("fit_is_empty"):
            return self
```

and in `fit_transform` call:
```python
     def fit_transform(self, X, y=None):
        X_inner, y_inner, metadata = self._fit_checks(X, y, False, True)
```

Reply:
Thanks! Looks all good to me now